### PR TITLE
Support CUDA stream with stream memory pool

### PR DESCRIPTION
**Pull request by sonots**
Fix https://github.com/cupy/cupy/issues/225

Support CUDA stream. `stream` can be specified with `with` statement or `use()` method as:

```python
import cupy as cp

x_gpu = cp.array([1, 2, 3])

with cp.cuda.stream.Stream():
   y_gpu = cp.linalg.norm(x_gpu)

stream = cp.cuda.stream.Stream()
stream.use()
y_gpu = cp.linalg.norm(x_gpu)
```

`nvprof --print-gpu-trace python ~/test_cupy.py` shows cupy kernels are executed in another stream.

To support CUDA stream, this PR changes memory pool, too.
A memory pool is created for **each stream** separately so that parallel computations among streams do not touch memory used in another stream.

Note that cudaMalloc is issued on default stream always.

## Incompatibility changes:

* `cupy.cuda.generator.RandomState`: `setStream()` method was removed
* `cupy.cuda.stream.Stream(null=True)` is prohibited to assure `cupy.cuda.stream.Stream.null` object is always used to specify the default stream.

### CODE REVIEW

**Comment by sonots** [comment]
Added implementation to support stream on all CUDA libraries, cudnn, cublas, curand, cusparse, cusolver, and thrust.

**Comment by sonots** [comment]
Because it is difficult to test where CUDA is really using stream, I added examples and verified by running them with `nvprof --print-gpu-trace`.

**Comment by sonots** [comment]
Wrote about incompatibility changes https://github.com/cupy/cupy/pull/306#issue-243605081

**Comment by sonots** [comment]
I was thinking to implement another version using cudaStreamAddCallback as:

```
setStream
kernel
streamAddCallback to release gpu memory (decrementing reference count to memory)
```

so that we can assure that gpu memory block is returned to memory pool after a kernel execution is finished.

However, based on my experiment, cudaStreamAddCallback drastically slowed kernel executions as https://gist.github.com/sonots/e98a95aaceae65a15d2b59a81befb023. 

So, I gave up the version, and now I think having separated memory pool for each stream is only way to support CUDA stream.

**Comment by sonots** [comment]
@okuta Do you have any other ideas?

**Comment by sonots** [comment]
Removed `[WIP]`

**Comment by sonots** [comment]
jenkins, test this please

**Comment by okuta** [comment]
This is good feature, thanks.
I have two concerns that are speed and memory consumption.
Do you have detail data?

**Comment by sonots** [comment]
## Test on default stream

cublas_default_stream.py

```python
import cupy
import time

device = cupy.cuda.device.Device(0)
memory_pool = cupy.cuda.MemoryPool()
cupy.cuda.set_allocator(memory_pool.malloc)

rand = cupy.random.generator.RandomState()

n = 10
start_time = time.time()
for i in range(n):
    x = rand.normal(size=(1, 1024 * 256))
    y = rand.normal(size=(1024 * 256, 1))
    z = cupy.matmul(x, y)
device.synchronize()
elapsed_time = time.time() - start_time
print('elapsed time', elapsed_time)
print('total bytes', memory_pool.total_bytes())
```

Before:

```
elapsed time 5.286596298217773
total bytes 6293504
```

After: 

```
elapsed time 5.262510299682617
total bytes 629350
```

## Test on a stream

cublas_stream.py

```python
import cupy
import time

device = cupy.cuda.device.Device(0)
memory_pool = cupy.cuda.MemoryPool()
cupy.cuda.set_allocator(memory_pool.malloc)

rand = cupy.random.generator.RandomState()
stream = cupy.cuda.stream.Stream()

n = 10
start_time = time.time()
for i in range(n):
    with stream:
        x = rand.normal(size=(1, 1024 * 256))
        y = rand.normal(size=(1024 * 256, 1))
        z = cupy.matmul(x, y)
device.synchronize()
elapsed_time = time.time() - start_time
print('elapsed time', elapsed_time)
print('total bytes', memory_pool.total_bytes())
```


After (only available with this patch):

```
elapsed time 5.287049865722656
total bytes 6293504
```

## Test on 10 streams

cublas_10streams.py

```python
import cupy
import time

device = cupy.cuda.device.Device(0)
memory_pool = cupy.cuda.MemoryPool()
cupy.cuda.set_allocator(memory_pool.malloc)
rand = cupy.random.generator.RandomState()

n = 10
start_time = time.time()
for i in range(n):
    stream = cupy.cuda.stream.Stream()
    with stream:
        x = rand.normal(size=(1, 1024 * 256))
        y = rand.normal(size=(1024 * 256, 1))
        z = cupy.matmul(x, y)
device.synchronize()
elapsed_time = time.time() - start_time
print('elapsed time', elapsed_time)
print('total bytes', memory_pool.total_bytes()
```

After (only available with this patch):

```
elapsed time 5.3070900440216064
total bytes 41963520
```

The reason that `total bytes` increases is because we have separated memory pool for each stream. The elapsed time increases a bit because we can not reuse memory pool among streams.

## Conclusion

There is almost no regression.
The increased memory consumption when we use multiple streams is as we expected

**Comment by sonots** [comment]
> BTW, looking nvvp, current kernel executions were not running on this example.

It was because `cudaMemset` was issued inside of `matmul`. Fixed via https://github.com/cupy/cupy/pull/306/commits/e518872e45667a3e7c763c888947c5cc71af6dc9

**Comment by okuta** [comment]
Please resolve conflicts.

**Comment by sonots** [comment]
rebased to master

**Comment by sonots** [comment]
rebased master, I will try to move `get_current_stream()` from python layer to cython (c) layer next.

**Comment by sonots** [comment]
To use unordered_map, I turned on `-std=c++11`.

**Comment by sonots** [comment]
I've created C API version of `get_current_stream_ptr()` as https://github.com/sonots/cupy/commit/499d2f442019b4968ecfb8af5923101d831b7399
https://github.com/sonots/cupy/tree/stream_pool2.

test.py

```
import cupy
import time

x = cupy.array([1])
y = cupy.array([1])

with cupy.cuda.stream.Stream():
    start = time.time()
    for i in range(0, 100):
        z = x + y
    elapsed_time = time.time() - start
    print ("elapsed_time:{0}".format(elapsed_time) + "[sec]")
```

Before:

```
$ python ~/test.py
elapsed_time:0.23342299461364746[sec]
$ python ~/test.py
elapsed_time:0.23365449905395508[sec]
```

```
$ perf record python ~/test.py
$ perf report
  18.50%  python    [kernel.kallsyms]                              [k] smp_call_function_many
  13.87%  python    [kernel.kallsyms]                              [k] clear_page_c_e
  13.87%  python    [kernel.kallsyms]                              [k] __switch_to
   9.83%  python    [kernel.kallsyms]                              [k] entry_SYSCALL_64
   8.67%  python    [kernel.kallsyms]                              [k] cascade
   7.51%  python    [kernel.kallsyms]                              [k] _raw_spin_lock
   6.94%  python    libopenblasp-r0-39a31c03.2.18.so               [.] blas_thread_server
   5.20%  bash      [kernel.kallsyms]                              [k] unmap_page_range
   5.20%  bash      [unknown]                                      [k] 0x00007fd3f0052f8b
   4.62%  bash      [kernel.kallsyms]                              [k] ___pmd_free_tlb
   4.62%  bash      [unknown]                                      [k] 0x000000000045a0cd
   0.58%  bash      [kernel.kallsyms]                              [k] flush_tlb_page
   0.58%  sed       [kernel.kallsyms]                              [k] copy_page
```

After (C API):

```
$ python ~/test.py
elapsed_time:0.23448991775512695[sec]
```

```
$ perf record python ~/test.py
$ perf report
  15.35%  python    [kernel.kallsyms]                    [.] entry_SYSCALL_64_fastpath
  14.01%  python    python3.6                            [.] PyUnicode_Compare
  12.01%  bash      [kernel.kallsyms]                    [k] __wake_up_bit
  11.34%  python    [kernel.kallsyms]                    [k] sched_clock_cpu
   8.01%  python    python3.6                            [.] _PyObject_Free
   7.34%  python    [kernel.kallsyms]                    [k] clear_buddies
   6.00%  python    [kernel.kallsyms]                    [k] memcpy_erms
   6.00%  pyenv-ve  [kernel.kallsyms]                    [k] __d_lookup_rcu
   5.34%  bash      [kernel.kallsyms]                    [k] __rmqueue.isra.80
   4.67%  bash      [kernel.kallsyms]                    [k] read_tsc
   4.67%  python    [kernel.kallsyms]                    [k] update_blocked_averages
   4.67%  bash      [kernel.kallsyms]                    [k] __compute_runnable_contrib
   0.58%  python    libc-2.23.so                         [.] malloc
```

**Comment by sonots** [comment]
I think implementation is fine in both C API and Python API of `get_current_stream()`, but we have to note that CPython API for Thread Local Storage will be changed in python 3.7. https://github.com/python/cpython/blob/731e18901484c75b60167a06a0ba0719a6d4827d/Include/pythread.h#L81-L92

**Comment by sonots** [comment]
Added the commit of C API `get_current_stream_ptr()`
Although the performance was not improved so much under my experiment, but I am sure that C API  version is faster than Python API version anyway.

**Comment by sonots** [comment]
hmm, it looks compilation fails with cuda9 ... ? 🤔 

```
21:36:29 cupy/cuda/cupy_thrust.cu(92): error: namespace "thrust::system::cuda::detail" has no member "execute_on_stream"
21:36:29 
21:36:29 cupy/cuda/cupy_thrust.cu(92): error: no instance of overloaded function "stable_sort" matches the argument list
21:36:29             argument types are: (thrust::cuda_cub::execute_on_stream, thrust::device_ptr<size_t>, thrust::device_ptr<size_t>, elem_less<cpy_byte>)
21:36:29           detected during instantiation of "void cupy::thrust::_lexsort<T>(size_t *, void *, size_t, size_t, size_t) [with T=cpy_byte]" 
21:36:29 (101): here
21:36:29 
21:36:29 cupy/cuda/cupy_thrust.cu(92): error: no instance of overloaded function "stable_sort" matches the argument list
21:36:29             argument types are: (thrust::cuda_cub::execute_on_stream, thrust::device_ptr<size_t>, thrust::device_ptr<size_t>, elem_less<cpy_ubyte>)
21:36:29           detected during instantiation of "void cupy::thrust::_lexsort<T>(size_t *, void *, size_t, size_t, size_t) [with T=cpy_ubyte]" 
21:36:29 (102): here
21:36:29 
21:36:29 cupy/cuda/cupy_thrust.cu(92): error: no instance of overloaded function "stable_sort" matches the argument list
21:36:29             argument types are: (thrust::cuda_cub::execute_on_stream, thrust::device_ptr<size_t>, thrust::device_ptr<size_t>, elem_less<cpy_short>)
21:36:29           detected during instantiation of "void cupy::thrust::_lexsort<T>(size_t *, void *, size_t, size_t, size_t) [with T=cpy_short]" 
21:36:29 (103): here
21:36:29 
```

**Comment by sonots** [comment]
jenkins, test this please

**Comment by sonots** [comment]
jenkins, test this please

**Comment by sonots** [comment]
doctest is failing on jenkins although it works on my machine ....

File: cupy/cuda/stream.pyx [hunk 1]
```diff
@@ -1,4 +1,48 @@
 from cupy.cuda import runtime
+from cpython cimport pythread
+import threading
+import weakref
+
+
+thread_local = threading.local()
```
Comment (okuta) [inline cupy/cuda/stream.pyx]: How about `cdef object _thread_local` ?

Comment (okuta) [inline cupy/cuda/stream.pyx]: How about `cdef object _thread_local` ?

File: cupy/cuda/stream.pyx [hunk 2]
```diff
@@ -1,4 +1,48 @@
...
+cdef int current_stream_key = pythread.PyThread_create_key()
```
Comment (okuta) [inline cupy/cuda/stream.pyx]: How about `_current_stream_key`?

**Comment by sonots** [comment]
I am now thinking to add an interface to `free_all_blocks` of an arena for a given stream.

**Comment by sonots** [comment]
Jenkins fails with `cd docs && make html` as:

```
/work/cupy/docs/source/reference/generated/cupy.cuda.Event.rst:53: WARNING: error while formatting arguments for cupy.cuda.Event.record: <cyfunction Event.record at 0x2aea1bcb8d10> is not a Python function
```

I am struggling to resolve this, but I have no clue how to resolve this issue.
I found pandas has same issue https://github.com/pandas-dev/pandas/issues/5218, but it seems they have also no idea.

**Comment by sonots** [comment]
Hmm, it seems `make html` succeeds with python 3.6.2, but fails with python 2.7.13.

**Comment by sonots** [comment]
jenkins, test this please

**Comment by sonots** [comment]
It seems `insepect.isfunction(cyfuction)` returns True from python 3.4 https://groups.google.com/forum/#!topic/cython-users/FvcMPk9n2X8. Can we use python >= 3.4 to generate our sphinx docs in chainer-test?

**Comment by sonots** [comment]
https://github.com/cupy/cupy/pull/306#issuecomment-338575431

I've debugged sphinx. sphinx has `config.suprress_warnings` http://www.sphinx-doc.org/ja/stable/config.html, but `cyfunction is not a Python function` was not supported.

Another way to avoid is not to treat warnings as error with a change:

Makefile

```
-  6 SPHINXBUILD   = sphinx-build -W
+   6 SPHINXBUILD   = sphinx-build
```

----

This is another problem, but I noticed that number of doctests were increased as:

sphinx-build -W

```
Doctest summary
===============
   60 tests
    0 failures in tests
    0 failures in setup code
    0 failures in cleanup code
```

sphinx-build

```
Doctest summary
===============
   95 tests
    0 failures in tests
    0 failures in setup code
    0 failures in cleanup code
```

**Comment by sonots** [comment]
`Memory.__init__ ` also get raised TypeError, but it seems it is not included in the warnings. It seems that it was reason why doctest did not fail before... 🤔 

```
('util/inspect.py:115', <cyfunction Memory.__init__ at 0x7f744a3cab90>)
('util/inspect.py:115', <cyfunction Stream.__init__ at 0x7f744afb6ad0>)
('util/inspect.py:115', <cyfunction Event.__init__ at 0x7f744afb6710>)
('util/inspect.py:115', <cyfunction ReductionKernel.__init__ at 0x7f744a3f64d0>)
('util/inspect.py:115', <cyfunction Event.__init__ at 0x7f744afb6710>)
('util/inspect.py:115', <cyfunction Event.record at 0x7f744afb6950>)
('ext/autodoc.py:751', <type 'exceptions.TypeError'>, TypeError('<cyfunction Event.record at 0x7f744afb6950> is not a Python function',))
('util/inspect.py:115', <cyfunction Event.synchronize at 0x7f744afb6a10>)
('ext/autodoc.py:751', <type 'exceptions.TypeError'>, TypeError('<cyfunction Event.synchronize at 0x7f744afb6a10> is not a Python function',))
('util/inspect.py:115', <cyfunction Memory.__init__ at 0x7f744a3cab90>)
('util/inspect.py:115', <cyfunction Stream.__init__ at 0x7f744afb6ad0>)
('util/inspect.py:115', <cyfunction Stream.__enter__ at 0x7f744afb6c50>)
('ext/autodoc.py:751', <type 'exceptions.TypeError'>, TypeError('<cyfunction Stream.__enter__ at 0x7f744afb6c50> is not a Python function',))
('util/inspect.py:115', <cyfunction Stream.__exit__ at 0x7f744afb6d10>)
('ext/autodoc.py:751', <type 'exceptions.TypeError'>, TypeError('<cyfunction Stream.__exit__ at 0x7f744afb6d10> is not a Python function',))
('util/inspect.py:115', <cyfunction Stream.add_callback at 0x7f744a3ca050>)
('ext/autodoc.py:751', <type 'exceptions.TypeError'>, TypeError('<cyfunction Stream.add_callback at 0x7f744a3ca050> is not a Python function',))
('util/inspect.py:115', <cyfunction Stream.record at 0x7f744a3ca110>)
('ext/autodoc.py:751', <type 'exceptions.TypeError'>, TypeError('<cyfunction Stream.record at 0x7f744a3ca110> is not a Python function',))
('util/inspect.py:115', <cyfunction Stream.synchronize at 0x7f744afb6f50>)
('ext/autodoc.py:751', <type 'exceptions.TypeError'>, TypeError('<cyfunction Stream.synchronize at 0x7f744afb6f50> is not a Python function',))
('util/inspect.py:115', <cyfunction Stream.use at 0x7f744afb6dd0>)
('ext/autodoc.py:751', <type 'exceptions.TypeError'>, TypeError('<cyfunction Stream.use at 0x7f744afb6dd0> is not a Python function',))
('util/inspect.py:115', <cyfunction Stream.wait_event at 0x7f744a3ca1d0>)
('ext/autodoc.py:751', <type 'exceptions.TypeError'>, TypeError('<cyfunction Stream.wait_event at 0x7f744a3ca1d0> is not a Python function',))
('util/inspect.py:115', <cyfunction ufunc.__init__ at 0x7f744a3e1b90>)
('util/inspect.py:115', <cyfunction ReductionKernel.__init__ at 0x7f744a3f64d0>)
('util/inspect.py:115', <cyfunction ufunc.__init__ at 0x7f744a3e1b90>)
```

**Comment by sonots** [comment]
jenkins, test this please

**Comment by sonots** [comment]
jenkins, test this please

**Comment by okuta** [comment]
LGTM!

**Comment by sunbearc22** [comment]
Can you provide a simple example of how to use Nvidia Visual profiler `nvvp` to profile a python-cupy script? Thanks.

**Pull request merged**